src/Orleans/Core/GrainStateStorageBridge.cs

```diff
@@ -125,13 +126,14 @@ public GrainStateStorageBridge(string grainTypeName, Grain grain, IStorageProvid
         {
             const string what = "ClearState";
             Stopwatch sw = Stopwatch.StartNew();
-            GrainReference grainRef = grain.GrainReference;
+            GrainReference grainRef = baseGrain.GrainReference;
             try
             {
                 // Clear (most likely Delete) state from external storage
                 await store.ClearStateAsync(grainTypeName, grainRef, grain.GrainState);
+
                 // Null out the in-memory copy of the state
-                grain.GrainState.SetAll(null);
+                grain.GrainState.State = null;
```

Comment:
I think this is a bug - instead of nulling `State` we need to set it to a fresh instance of the state class.

`Grain_AzureStore_Delete` test is failing because of this. It tries to read a value of a state property right after calling `ClearStateAsync`, and the grain code throws a `NullReferenceException`.